### PR TITLE
#167729791 Profile Page Set Up

### PR DESCRIPTION
**Pull request by chibytez**
#### What does this PR do?
Set up  user profile page
#### Description of Task to be completed?
added get profile controller
added update profile controller
#### How should this be manually tested?
Can be tested on postman by visiting these routes
localhost:3000/api/v1/users/profile with a GET and 
PUT request by adding these fields:
 gender, birthday, preferredlanguage,  currency ,residentialaddress, role, department, linemanager

#### Any background context you want to provide?
none
#### What are the relevant pivotal tracker stories?
#167729791
#### Screenshots (if appropriate)
Capture
Capture
Capture
#### Questions:

### CODE REVIEW

**Review by mthamayor — CHANGES_REQUESTED** [review]
https://github.com/andela/shadowcat-backend/blob/a6100e4ba7fda74d91de2bd2474ba0cf7ff30be3/src/utils/errorMessage.js#L12
For this line, can you please check the signup password requirements and update accordingly?

**Review by mthamayor — CHANGES_REQUESTED** [review]
https://github.com/andela/shadowcat-backend/blob/a6100e4ba7fda74d91de2bd2474ba0cf7ff30be3/src/utils/errorMessage.js#L16
I think it should be `Birthday should be Alphanumeric`

**Review by mthamayor — CHANGES_REQUESTED** [review]
https://github.com/andela/shadowcat-backend/blob/4d1366e41e6a55d1aa86d2012f2c4138a0880ada/src/utils/constants/errorMessages.js#L51

There should be only one `default export` and please make your `export const Errors` more descriptive. something like `export const profileErrors`

**Comment by chibytez** [comment]
> https://github.com/andela/shadowcat-backend/blob/4d1366e41e6a55d1aa86d2012f2c4138a0880ada/src/utils/constants/errorMessages.js#L51
> 
> There should be only one `default export` and please make your `export const Errors` more descriptive. something like `export const profileErrors`

Errors take care of both login and profile errors

**Review by mthamayor — APPROVED** [review]
Thumbs up.

**Review by sebaztyn — APPROVED** [review]
Good work bro

File: src/database/migrations/20190901120620-adding_to_user.js [hunk 1]
```diff
@@ -0,0 +1,46 @@
+export default {
+  up: async (queryInterface, Sequelize) => {
+    await queryInterface.addColumn('Users', 'gender', {
+      type: Sequelize.STRING,
+      allowNull: true
+    });
+    await queryInterface.addColumn('Users', 'birthday', {
+      type: Sequelize.STRING,
```
Comment (Stylll) [inline src/database/migrations/20190901120620-adding_to_user.js]: The type of birthday should be `DATE` not `STRING`

File: src/database/migrations/20190901120620-adding_to_user.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    await queryInterface.addColumn('Users', 'currency', {
+      type: Sequelize.STRING,
+      allowNull: true
+    });
+    await queryInterface.addColumn('Users', 'residentialaddress', {
+      type: Sequelize.STRING,
+      allowNull: true
+    });
+    await queryInterface.addColumn('Users', 'role', {
+      type: Sequelize.STRING,
```
Comment (Stylll) [inline src/database/migrations/20190901120620-adding_to_user.js:24]: User roles should have an enum type. I'm not certain how many roles we can have but you can reach out to @sebaztyn on creating an enum type and setting the values for it

File: src/database/migrations/20190901120620-adding_to_user.js [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+      allowNull: true
+    });
+    await queryInterface.addColumn('Users', 'department', {
+      type: Sequelize.STRING,
+      allowNull: true
+    });
+    await queryInterface.addColumn('Users', 'linemanager', {
+      type: Sequelize.STRING,
```
Comment (Stylll) [inline src/database/migrations/20190901120620-adding_to_user.js]: The linemanager is going to be a user in the app too so we should save the linemanager's user id and not the name. You will need to change the type here to INTEGER

File: src/middlewares/auth.js [hunk 4]
```diff
@@ -67,7 +67,10 @@ class Authentication {
    */
   static decodeJwt(token) {
     let payload = null;
-    payload = jwt.decode(token, index.secret);
+    payload = jwt.verify(token, index.secret, (err, decoded) =>{
```
Comment (Stylll) [inline src/middlewares/auth.js]: is there a reason we are changing the existing line to this?

File: src/routes/api/index.js [hunk 5]
```diff
@@ -1,15 +1,12 @@
...
 
 const usersRouter = express.Router();
 
 usersRouter.use('/v1/auth', authRouter);
-
-usersRouter.use('/v1/users', passwordRouter);
-
-usersRouter.get('/', (req, res) => res.status(200).send('Welcome to  Shadowcat API'));
-
+usersRouter.use('/v1/profiles', profileRoute);
```
Comment (Stylll) [inline src/routes/api/index.js]: `profiles` is a route that belongs to a `user` so this should be under the `user` route like `/v1/users/profiles`

File: src/routes/api/profile.js [hunk 6]
```diff
@@ -0,0 +1,11 @@
+import { Router } from 'express';
+import ProfileController from '../../controllers/profileController';
+import Authentication from '../../middlewares/auth';
+import profileValidation from '../../validation/profilevalidation';
+
+const profileRoute = Router();
+
+profileRoute.get('/', Authentication.authenticate, ProfileController.getProfile);
+profileRoute.put('/', Authentication.authenticate, profileValidation, ProfileController.updateProfile);
```
Comment (Stylll) [inline src/routes/api/profile.js]: These routes should be under the user route

File: src/controllers/profileController.js [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+      const { id } = req;
+      const profileDetails = await (req.body);
+      const {
+        firstname, lastname, email, gmail, facebook, gender, birthday,
+        preferredlanguage, currency, residentialaddress, role, department, linemanager
+      } = profileDetails;
+      const user = await Users.findOne({
+        where: { userId: id }
+      });
+      if (user !== undefined) {
```
Comment (Stylll) [inline src/controllers/profileController.js:89]: I see we are handling the case where user exists, we should also return a `user not found` response if the user does not exist.

**Review by Stylll — CHANGES_REQUESTED** [review]
Hey @chibytez, good work so far. I left a few comments. Please review them.

File: src/middlewares/auth.js (hunk 4 above)
Comment (chibytez) [reply src/middlewares/auth.js]: Yeah we want to verify the token which jwt.verify is what is well suited for it

File: src/routes/api/index.js (hunk 5 above)
Comment (chibytez) [reply src/routes/api/index.js]: ok thanks a lot

File: src/controllers/profileController.js (hunk 7 above)
Comment (chibytez) [reply src/controllers/profileController.js:89]: ok will add that

File: src/database/migrations/20190901120620-adding_to_user.js [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+        'Maintenance',
+      ],
+      allowNull: true
+    });
+    await queryInterface.addColumn('Users', 'department', {
+      type: Sequelize.STRING,
+      allowNull: true
+    });
+    await queryInterface.addColumn('Users', 'linemanager', {
+      type: Sequelize.INTEGER,
```
Comment (sebaztyn) [inline src/database/migrations/20190901120620-adding_to_user.js]: your line manager column should reference a userId on your table as per @Stylll advice

File: src/models/user.js [hunk 9]
```diff
@@ -42,7 +42,44 @@ module.exports = (sequelize, DataTypes) => {
...
+          'Maintenance',
+        ],
+        required: false
+      },
+      department: {
+        type: DataTypes.STRING,
+        required: false
+      },
+      linemanager: {
+        type: DataTypes.INTEGER,
```
Comment (sebaztyn) [inline src/models/user.js]: Please take a look at your line manager's data type to allow for referencing to a data type corresponding to UUID which saves the userId.

**Review by sebaztyn — APPROVED** [review]

**Review by mthamayor — APPROVED** [review]

File: src/routes/api/index.js [hunk 10]
```diff
@@ -1,15 +1,12 @@
...
 
 const usersRouter = express.Router();
 
 usersRouter.use('/v1/auth', authRouter);
-
-usersRouter.use('/v1/users', passwordRouter);
-
-usersRouter.get('/', (req, res) => res.status(200).send('Welcome to  Shadowcat API'));
-
+usersRouter.use('/v1/users/profiles', profileRoute);
```
Comment (Stylll) [inline src/routes/api/index.js]: this should be `/v1/users/profile` not `/v1/users/profiles`

File: src/utils/constants/errorMessages.js [hunk 11]
```diff
@@ -19,4 +19,33 @@ export const signupVerifyErrors = {
   notFound: 'No pending verification found'
 };
 
+export const loginProfileErrors = {
```
Comment (Stylll) [inline src/utils/constants/errorMessages.js]: this object name can be `userProfileErrors`. That's a bit more specific to users that `loginProfileErrors`

File: src/utils/constants/errorMessages.js [hunk 12]
```diff
@@ -19,4 +19,33 @@ export const signupVerifyErrors = {
...
+  undefinedEmail: 'Email is required',
+  undefinedPassword: 'Password is required',
+  invalidPassword: 'Password cannot be less than 8 characters',
+  invalidEmail: 'Enter a valid email address',
+  nonAndelanEmail: 'Email should be an andela email',
+  alphaNumericPassword: 'Password should be alphanumeric',
+  undefinedGender: 'Gender is required',
+  invalidGender: 'Gender should be an Alphabet',
+  undefinedBirthday: 'Birthday is required',
+  invalidBirthday: 'Birthday should be  a date',
```
Comment (Stylll) [inline src/utils/constants/errorMessages.js]: Please check through the whole file and remove double spacing in the strings

File: src/utils/helpers/pullErrors.js [hunk 13]
```diff
@@ -50,6 +82,30 @@ const pullErrors = (errors) => {
...
     combinedErrors.password = password;
   }
+  if (gender.length > 0) {
+    combinedErrors.gender = gender;
+  }
+  if (birthday.length > 0) {
+    combinedErrors.birthday = birthday;
+  }
+  if (preferredlanguage.length > 0) {
+    combinedErrors.preferrdlanguage = preferredlanguage;
```
Comment (Stylll) [inline src/utils/helpers/pullErrors.js]: what does `preferrdlanguage` mean?. I think it should be `preferredlanguage` as that's more readable.

File: src/utils/helpers/pullErrors.js [hunk 14]
```diff
@@ -10,6 +10,14 @@ const pullErrors = (errors) => {
   const email = [];
   const phone = [];
   const password = [];
+  const gender = [];
+  const birthday = [];
+  const preferredlanguage = [];
+  const currency = [];
+  const residentialaddress = [];
```
Comment (Stylll) [inline src/utils/helpers/pullErrors.js]: Please check through this file to make sure all variable names are `camelCased` as that's the convention agreed on by the team

**Review by Stylll — CHANGES_REQUESTED** [review]
Hey @chibytez, I left a few comments. Please check them out.

**Review by davidfrank96 — APPROVED** [review]
Looks Good to me ... well done

**Review by mthamayor — APPROVED** [review]

File: src/middlewares/verification.js [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+        const user = await Users.findOne({
+          where: {
+            id: linemanager
+          }
+        });
+        if (!user) {
+          return response.sendError(
+            res,
+            404,
+            'line manager does not exit',
```
Comment (Stylll) [inline src/middlewares/verification.js]: This should be `Line manager does not exist`.

File: src/utils/constants/errorMessages.js [hunk 16]
```diff
@@ -19,4 +19,33 @@ export const signupVerifyErrors = {
...
+  nonAndelanEmail: 'Email should be an andela email',
+  alphaNumericPassword: 'Password should be alphanumeric',
+  undefinedGender: 'Gender is required',
+  invalidGender: 'Gender should be either male or female',
+  undefinedBirthday: 'Birthday is required',
+  invalidBirthday: 'Birthday should be a date format mm-dd-yyyy',
+  undefinedPreferredLanguage: 'Preferred Language is required',
+  invalidPreferredLanguage: 'Preferred Language should be an Alphabet with no spaces',
+  undefinedCurrency: 'Currency is required',
+  invalidCurrency: 'Currency should be an Alphabet with no spaces',
```
Comment (Stylll) [inline src/utils/constants/errorMessages.js]: Technically, an Alphabet is one letter. I think we should change this message and other similar ones to something like `Currency should be a set of alphabets with no spaces`.

File: src/utils/constants/errorMessages.js [hunk 17]
```diff
@@ -19,4 +19,33 @@ export const signupVerifyErrors = {
...
+  undefinedResidentialAddress: 'Residential address is required',
+  invalidResidentialAddress: 'Residential address should be an address',
+  undefinedRole: 'role is required',
+  invalidRole: 'role can be developer,qa or maintenance',
+  undefinedDepartment: 'Department is required',
+  invalidDepartment: 'Department should be an aqlphabet with no spaces',
```
Comment (Stylll) [inline src/utils/constants/errorMessages.js]: spell check here `alphabet` not `aqlphabet`.

**Review by Stylll — CHANGES_REQUESTED** [review]
Hey @chibytez , I left a few comments. I also noticed this while testing:
when I pass this request object
```
{
	"gender": "Animal",
	"birthday": "2010",
	"preferredlanguage": "German",
	"currency": "Pounds and Dollarz",
	"residentialaddress": "Moon and solar system",
	"role": "qae",
	"department": "fire squad",
	"linemanager": "4"
}
```
the error message looks like this
```
{
    "status": 400,
    "error": {
        "currency": [
            "Currency should be an Alphabet with no spaces"
        ],
        "role": [
            "Gender should be either male or female",
            "role can be developer,qa or maintenance"
        ],
        "department": [
            "Department should be an aqlphabet with no spaces"
        ]
    }
}
```
The error message for gender is placed inside the role array. Please fix that.

**Review by davidfrank96 — APPROVED** [review]
He has implemented the feedbacks

**Review by mthamayor — CHANGES_REQUESTED** [review]
```
{
	"gender": "Animal",
	"birthday": "2010",
	"preferredlanguage": "German",
	"currency": "Pounds and Dollarz",
	"residentialaddress": "Moon and solar system",
	"role": "qae",
	"department": "fire squad",
	"linemanager": "4s"
}
```

When I pass this request body(edited linemanager), I get a sequelize error:
`SequelizeDatabaseError: invalid input syntax for integer`
Can you take a look at this error?
Does it show up with you?

**Review by mthamayor — CHANGES_REQUESTED** [review]
```
{
	"gender": "male",
	"birthday": "10-13-2010",
	"preferredlanguage": "German",
	"currency": "Pounds",
	"residentialaddress": "Moon and solar system",
	"role": "developer",
	"department": "fire",
	"linemanager": "4"
}
```
I get this response even after passing `developer` as a user role
```
{
    "status": 400,
    "error": {
        "role": [
            "role can be developer,qa or maintenance"
        ]
    }
}
```
Can you try to break the code more and fix it appropriately before reaching out?  You would need to thoroughly test your app before asking for reviews

**Review by sebaztyn — APPROVED** [review]
Great work bro

**Review by Stylll — APPROVED** [review]

**Pull request merged**